interface/ceed-tensor.c

```diff
@@ -0,0 +1,165 @@
+///
+/// @addtogroup CeedBasis
+/// @{
+
+/**
+  @brief Create a CeedTensorContract object for a CeedBasis
+
+  @param ceed       A Ceed object where the CeedTensorContract will be created
+  @param contract   Address of the variable where the newly created
+                      CeedTensorContract will be stored.
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedTensorContractCreate(Ceed ceed, CeedTensorContract *contract) {
+  int ierr;
+
+  if (!ceed->TensorContractCreate) {
+    Ceed delegate;
+    ierr = CeedGetDelegate(ceed, &delegate); CeedChk(ierr);
+
+    if (!delegate)
+      return CeedError(ceed, 1,
+                       "Backend does not support TensorContractCreate");
+
+    ierr = CeedTensorContractCreate(delegate, contract);
+    CeedChk(ierr);
+    return 0;
+  }
+
+  ierr = CeedCalloc(1,contract); CeedChk(ierr);
+
+  (*contract)->ceed = ceed;
+  ceed->refcount++;
+  ierr = ceed->TensorContractCreate(*contract);
+  CeedChk(ierr);
+  return 0;
+};
+
+/**
+  @brief Apply tensor contraction
+
+    Contracts on the middle index
+    NOTRANSPOSE: v_ajc = t_jb u_abc
+    TRANSPOSE:   v_ajc = t_bj u_abc
+    If add != 0, "=" is replaced by "+="
+
+  @param contract   CeedTensorContract to use
+  @param A          First index of u, v
+  @param B          Middle index of u, one index of t
+  @param C          Last index of u, v
+  @param J          Middle index of v, one index of t
+  @param[in] t      Tensor array to contract against
+  @param tmode      Transpose mode for t, \ref CEED_NOTRANSPOSE for t_jb
+                    \ref CEED_TRANSPOSE for t_bj
+  @param add        Add mode
+  @param[in] u      Input array
+  @param[out] v     Output array
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedTensorContractApply(CeedTensorContract contract, CeedInt A, CeedInt B,
+                            CeedInt C, CeedInt J, const CeedScalar *restrict t,
+                            CeedTransposeMode tmode, const CeedInt add,
+                            const CeedScalar *restrict u,
+                            CeedScalar *restrict v) {
+  int ierr;
+
+  ierr = contract->Apply(contract, A, B, C, J, t, tmode, add,  u, v);
+  CeedChk(ierr);
+  return 0;
+};
+
+/**
+  @brief Get Ceed associated with a CeedTensorContract
+
+  @param contract      CeedTensorContract
+  @param[out] ceed  Variable to store Ceed
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedTensorContractGetCeed(CeedTensorContract contract, Ceed *ceed) {
+  *ceed = contract->ceed;
+
+  return 0;
+};
+
+/**
+  @brief Get backend data of a CeedTensorContract
+
+  @param contract   CeedTensorContract
+  @param[out] data  Variable to store data
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedTensorContractGetData(CeedTensorContract contract, void* *data) {
```

Comment:
A TensorContract may require specialized persistent data. I included this although it is currently unneeded and untested.